tests/Unit/Evolution/Systems/ScalarWave/Test_Constraints.cpp

```diff
@@ -0,0 +1,351 @@
+      static_cast<tnsr::ij<DataVector, Dim, Frame::Inertial> (*)(
+          const tnsr::ij<DataVector, Dim, Frame::Inertial>&)>(
+          &ScalarWave::two_index_constraint<Dim>),
+      "TestFunctions", "two_index_constraint", {{{-10.0, 10.0}}}, used_for_size,
+      1.0e-12);
+}
+
+// Test the return-by-reference two-index constraint
+// by comparing to a time-independent analytic solution (e.g. Kerr-Schild)
+template <typename Solution>
+void test_two_index_constraint_analytic(
+    const Solution& solution, const size_t grid_size_each_dimension,
+    const std::array<double, 3>& lower_bound,
+    const std::array<double, 3>& upper_bound,
+    const double error_tolerance) noexcept {
+  // Shorter names for tags.
+  constexpr size_t dim = 3;
+  using VariablesTags =
+      tmpl::list<ScalarWave::Pi, ScalarWave::Phi<dim>, ScalarWave::Psi>;
+
+  // Check vs. time-independent analytic solution
+  // Set up grid
+  const size_t data_size = pow<3>(grid_size_each_dimension);
+  Mesh<3> mesh{grid_size_each_dimension, Spectral::Basis::Legendre,
+               Spectral::Quadrature::GaussLobatto};
+
+  using Affine = domain::CoordinateMaps::Affine;
+  using Affine3D =
+      domain::CoordinateMaps::ProductOf3Maps<Affine, Affine, Affine>;
+  const auto coord_map =
+      domain::make_coordinate_map<Frame::Logical, Frame::Inertial>(Affine3D{
+          Affine{-1.0, 1.0, lower_bound[0], upper_bound[0]},
+          Affine{-1.0, 1.0, lower_bound[1], upper_bound[1]},
+          Affine{-1.0, 1.0, lower_bound[2], upper_bound[2]},
+      });
+
+  // Set up coordinates
+  const auto x_logical = logical_coordinates(mesh);
+  const auto x = coord_map(x_logical);
+  // Arbitrary time for time-independent solution.
+  const double t = 0.;
+
+  // Evaluate analytic solution
+  const auto vars = solution.variables(
+      x, t,
+      tmpl::list<ScalarWave::Pi, ScalarWave::Phi<dim>, ScalarWave::Psi>{});
+
+  // Compute derivatives d_psi, d_phi, d_pi, and d_gauge_function numerically
+  Variables<VariablesTags> sw_vars(data_size);
+  get<ScalarWave::Psi>(sw_vars) = get<ScalarWave::Psi>(vars);
+  get<ScalarWave::Pi>(sw_vars) = get<ScalarWave::Pi>(vars);
+  get<ScalarWave::Phi<dim>>(sw_vars) = get<ScalarWave::Phi<dim>>(vars);
+
+  // Compute numerical derivatives of psi,pi,phi,H
+  const auto sw_derivs =
+      partial_derivatives<VariablesTags, VariablesTags, dim, Frame::Inertial>(
+          sw_vars, mesh, coord_map.inv_jacobian(x_logical));
+  const auto& d_phi = get<
+      Tags::deriv<ScalarWave::Phi<dim>, tmpl::size_t<dim>, Frame::Inertial>>(
+      sw_derivs);
+
+  // Get the constraint, and check that it vanishes to error_tolerance
+  auto two_index_constraint =
+      make_with_value<tnsr::ij<DataVector, dim, Frame::Inertial>>(
+          x, std::numeric_limits<double>::signaling_NaN());
+  ScalarWave::two_index_constraint(make_not_null(&two_index_constraint), d_phi);
+
+  Approx numerical_approx =
+      Approx::custom().epsilon(error_tolerance).scale(1.0);
```

Comment:
The `scale(1.0)` means that there is an **absolute** tolerance of `error_tolerance` as well as a relative tolerance. Is that what you want?

Reply:
As the variable being checked should be 0 here, I think we do need an absolute tolerance. A relative tolerance alone, multiplied by the correct value, will give an allowed range of [0,0] for the computed answer.